[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52102 [JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization**

Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

desertfire committed:
[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1559,3 +1560,33 @@ def forward(self, x):
+
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    def test_freeze_conv_relu_fusion(self):
+        class Net(nn.Module):
```
We're going to want to add more tests around bias being present/not present, and NCHW memory format, as well as Conv1d/Conv2d/Conv3d. This is a good example test: https://github.com/pytorch/pytorch/blob/master/test/jit/test_freezing.py#L1373.

eellison reviewed (commented):
Great start 🚀 🚀 🚀 

I would say the big things we should figure out before landing is:
-  bias not present
- interaction with `torch.backends.cudnn.benchmark` - in the first version of this PR, just doing the default algorithm is fine, and we can integrate with the flag in a second PR
- We need to audit the scenarios under which pytorch eager [doesn't dispatch to cudnn](https://github.com/pytorch/pytorch/blob/0a8058469e1cd3642ab5a275512d8ed63cf163a1/aten/src/ATen/native/Convolution.cpp#L190). There may be cases where cudnn does not correctly compute things. Those cases that depend on the weight tensor we can avoid AOT, but we might need to do some runtime logic that depends on the input tensors.

eellison commented on aten/src/ATen/cudnn/Descriptors.h:
```diff
@@ -301,6 +301,23 @@ struct TORCH_CUDA_CPP_API CTCLossDescriptor
 #endif
 };
 
+struct TORCH_CUDA_CPP_API ActivationDescriptor
```
Cool!! We should be able to use this in the future to support other activations as well https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnActivationMode_t.

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire commented on aten/src/ATen/native/cudnn/Conv_v7.cpp:
```diff
@@ -821,6 +821,98 @@ void raw_cudnn_convolution_backward_weight_out(
+  bdesc.set(bias.expand({1, bias.size(0)}), output.dim());
+
+  ActivationDescriptor adesc;
+  adesc.set(CUDNN_ACTIVATION_RELU);
```
By calling "AlgoIterator<cudnnConvolutionFwdAlgoPerf_t>(args, benchmark)", we are leveraging https://github.com/pytorch/pytorch/blob/75cba9d0d18752bb800a1a7697976536590ccf5b/aten/src/ATen/native/cudnn/Conv_v7.cpp#L280 to fetch the best conv algorithm.

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire commented on aten/src/ATen/native/cudnn/Conv_v7.cpp:
```diff
@@ -821,6 +821,98 @@ void raw_cudnn_convolution_backward_weight_out(
+                args.cdesc.mut_desc(), fwdAlgPerf.mathType),
+            args);
+
+        Constant one(dataType, 1);
```
CUDNN API doc says this is the full statement, y = act ( alpha1 * conv(x) + alpha2 * z + bias ); In our implementation, alpha1 is always 1, and alpha2 is the passed-in alpha parameter.

eellison commented on aten/src/ATen/native/cudnn/Conv_v7.cpp:
```diff
@@ -821,6 +821,109 @@ void raw_cudnn_convolution_backward_weight_out(
+  ActivationDescriptor adesc;
+  adesc.set(CUDNN_ACTIVATION_RELU);
+
+  AlgoIterator<cudnnConvolutionFwdAlgoPerf_t>(args, benchmark)
```
The equivalent pytorch eager code when we're not benchmarking is just:
```
    if (!benchmark) {
      AT_CUDNN_CHECK_WITH_SHAPES(cudnnGetConvolutionForwardAlgorithm_v7(
          args.handle,
          args.idesc.desc(),
          args.wdesc.desc(),
          args.cdesc.desc(),
          args.odesc.desc(),
          num_algos,
          &perf_count,
          perf_results.get()), args);
    } else { 
```

why are we doing the `try_all([&](const cudnnConvolutionFwdAlgoPerf_t` logic ?

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1730,3 +1730,50 @@ def test_conv_to_mkldnn_no_mkldnn(self):
+                    self.add_z = add_z
+
+                def forward(self, x):
+                    identity = x
```
i would consider making `identity = torch.rand_like(x)` just to stress this a little more

eellison commented on torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp:
```diff
@@ -0,0 +1,80 @@
+  // matching
+  graph_rewrite_helper::replaceConvolutionWithAtenConv(graph);
+  RemoveTensorMutation(graph, [](Node* node) {
+    return node->kind() == aten::add_ || node->kind() == aten::relu_;
```
we might consider making this so we only remove them if an input to add_ is Convolution and an input to relu_ is Convolution or an add_ which has a convolution input. (not necessary for the first PR)

eellison replied:
for (Value * v: node->inputs()) {
    if (v->node()->kind() == aten::conv2d) {
    
    }
}

eellison commented on torch/csrc/jit/passes/frozen_graph_optimizations.cpp:
```diff
@@ -21,6 +25,8 @@ void OptimizeFrozenGraph(
     }
   }
+
+  FuseFrozenConvRelu(graph);
```
I think we might want to integrate this into the main api only after a little more testing on benchmark suite and such.

desertfire replied:
Let me revert this one.

eellison commented on torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp:
```diff
@@ -0,0 +1,80 @@
+
+  auto is_cuda = [](const Match& match,
+                    const std::unordered_map<std::string, Value*>& vmap) {
+    // auto input = toIValue(match.values_map.at(vmap.at("input"))).value();
```
leftover code

eellison commented on aten/src/ATen/native/cudnn/ConvShared.cpp:
```diff
@@ -491,6 +491,77 @@ Tensor cudnn_convolution_transpose_backward_weight(
+    IntArrayRef dilation,
+    int64_t groups) {
+  // Skip all the shape and type checking because frozen-model opt should
+  // catch problems during the compilation
```
We should actually add this logic to the fusion pass bc it doesnt exist right now

eellison commented on torch/csrc/jit/passes/frozen_conv_relu_fusion.cpp:
```diff
@@ -0,0 +1,80 @@
+    // if (input.toTensor().suggest_memory_format() !=
+    // at::MemoryFormat::ChannelsLast)
+    //  return false;
+    auto weight = toIValue(match.values_map.at(vmap.at("weight"))).value();
```
if bias or `z` are present we might consider doing some error checking up front.

eellison commented on aten/src/ATen/native/cudnn/Conv_v7.cpp:
```diff
@@ -821,6 +821,109 @@ void raw_cudnn_convolution_backward_weight_out(
+    int64_t groups) {
+  // This function is currently only called if frozen-model opt is turned on.
+  // The benchmarking and caching mechanism is not ready, so let's set benchmark
+  // to false.
```
https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnActivationMode_t : ` cudnnConvolutionBiasActivationForward() function must use CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_​PRECOMP_GEMM` - i'm not sure we can implement benchmarking at all.

desertfire replied:
"Only the CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_​PRECOMP_GEMM algo is enabled with CUDNN_ACTIVATION_IDENTITY. In other words, in the cudnnActivationDescriptor_t structure of the input activationDesc, if the mode of the cudnnActivationMode_t field is set to the enum value CUDNN_ACTIVATION_IDENTITY, then the input cudnnConvolutionFwdAlgo_t of this function cudnnConvolutionBiasActivationForward() must be set to the enum value CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_​PRECOMP_GEMM. For more information, see cudnnSetActivationDescriptor()."  -- My interpretation is we can still do benchmarking if the activation function is RELU.

eellison reviewed (approved):
Looks great 😍 😍 😍 

I think we might consider turning it on in freezing as a separate PR just so this is less likely to break anything, and we can do a little more testing/benchmarking on a variety of models.

It would great if you would post a few benchmarks onto the pr description - i know you said up to around 3% when it was just Conv-Relu, not yet Conv-Add-Relu, i'd be curious about the latter case. Considering Conv-Relu and Conv-Add-Relu is like the entire compute pattern of many vision models after freezing, so we should definitely take whatever we can get.

It might be unlikely but we can keep our eyes out in the future for Convolution-Add - (Non pointwise op) and consider fusing with `CUDNN_ACTIVATION_IDENTITY`.

Tagging @ngimel for someone else familiar with Cudnn, would be great to get another set of eyes on it.

ngimel commented on aten/src/ATen/native/cudnn/ConvShared.cpp:
```diff
@@ -491,6 +491,77 @@ Tensor cudnn_convolution_transpose_backward_weight(
+    bias_contig = bias->contiguous(memory_format);
+    bias_contig.resize_(bias_contig.sizes(), memory_format);
+  } else {
+    bias_contig = *output;
```
you are setting bias to uninitialized output tensor here, how does this work?

desertfire replied:
I thought "auto output_t = at::native::empty_cuda(..." initializes output as zero. If it is not the right assumption, I will create zeros explicitly.

ngimel replied:
It's not the right assumption, but ideally you should avoid explicitly creating tensor of zeros, and instead use corresponding arguments in cudnn call (like beta=0) to make sure those uninitialized values are never read and only written to.

eellison commented on test/jit/test_freezing.py:
```diff
@@ -1730,3 +1730,50 @@ def test_conv_to_mkldnn_no_mkldnn(self):
+            self.run_pass("inline", scripted_mod.graph)
+            FileCheck().check("aten::relu").run(scripted_mod.graph)
+            frozen_mod = torch.jit.freeze(scripted_mod)
+            self.run_pass("fuse_frozen_conv_relu", frozen_mod.graph)
```
maybe add a test that checks if torch.jit.save will error (we dont want to have to maintain BC for this operator)

desertfire commented on aten/src/ATen/native/cudnn/ConvShared.cpp:
```diff
@@ -491,6 +491,77 @@ Tensor cudnn_convolution_transpose_backward_weight(
+  if (bias_t.has_value()) {
+    TensorArg bias{bias_t.value(), "bias", 3};
+    bias_contig = bias->contiguous(memory_format);
+    bias_contig.resize_(bias_contig.sizes(), memory_format);
```
I thought "auto output_t = at::native::empty_cuda(..." initializes output as zero. If it is not the right assumption, I will create zeros explicitly.

desertfire committed:
Update on "[JIT] Add CUDA Conv-Relu fusion for Freezing Model Optmization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Relu fusion for Frozen Model Optmization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Add CUDNN Conv-Add-Relu fusion for Frozen Model Optimization"


Summary: Fuse Conv-Relu based on cudnnConvolutionBiasActivationForward as a part of the Frozen Model Optimization. The optimization is not turned on as default yet.

Test Plan:
```
python test/test_jit.py -k test_freeze_conv_relu_fusion
```

Reviewers:

Subscribers:

Tasks: T78068152

Tags:

Differential Revision: [D26646100](https://our.internmc.facebook.com/intern/diff/D26646100)

[ghstack-poisoned]
